File: algorithms/0876.middle-of-the-linked-list/solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def middle_node(head)
  tmp, i = [], 0
  while (head)
    tmp[i] = head.val
    i += 1
    head = head.next
  end
  tmp[i / 2..-1]
end

File: algorithms/0876.middle-of-the-linked-list/solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut len = 0usize;
        let mut tmp = head.as_ref();
        while let Some(node) = tmp.as_ref() {
            len += 1;
            tmp = node.next.as_ref();
        }

        let mut head = head;
        for _ in 0..len / 2 {
            if let Some(node) = head {
                head = node.next;
            }
        }

        head
    }
}
